RFC-0005-multi-cluster-resource-management.md

```diff
@@ -0,0 +1,64 @@
+
+The idea of moving the queueing phase out of the coordinator into a separate process is an old one dating back many years. Here are a few previous proposals/discussions: 
+* https://github.com/prestodb/presto/pull/12176
+* https://github.com/trinodb/trino/issues/391
+* https://github.com/prestodb/presto/issues/10174
+* https://github.com/prestodb/presto/issues/13198
+
+## Background
+
+Currently, each Presto cluster owns its own resource management and queue. The resource manager of each cluster decides which query in the cluster’s queue should be executed next, based on logic described in the cluster’s [resource group configuration](https://prestodb.io/docs/current/admin/resource-groups.html). 
+
+This status quo of cluster-level queueing works fine to manage one or a few clusters, but quickly becomes problematic when managing a deployment of tens or hundreds of Presto clusters. Some issues include: 
+
+1. Due to the queue being fragmented across multiple clusters, and the admission decision being made at the cluster level, resource group admission decisions are not correct globally, although they could be correct locally. For example, it is possible that one cluster might admit a query of lower priority than a query in another cluster’s queue. 
+2. Since query concurrency control is applied on a per-cluster basis, global concurrency control can only be achieved, if the concurrency limit is a multiple of the number of clusters. For example, it is currently impossible to enforce a resource group with a concurrency limit of 1 query, across multiple clusters. 
+3. Load-balancing queries effectively across multiple clusters requires the load balancer to understand the resource group and queue state of each cluster. This is difficult to achieve, and in practice, stale information leads to poor load balancing. 
+4. Since the query is queued on the cluster itself, restarting clusters for maintenance reasons requires either that the queued queries be killed, or a lengthy drain time. 
+
+## Proposed Implementation
+
+We propose to add support in Presto for an optional queueing service that sits upstream of multiple Presto clusters. This queueing service will serve as the global entry-point for all queries from the client and will implement the QueuedStatement protocol. This queueing service is responsible for: 
+
+1. Authenticating the request 
+2. Adding the query to the global queue 
+3. Responding to clients polling for the queueing status of their query
+4. Picking the next query to execute 
+5. Selecting the best cluster for execution of this query
+6. Dispatching this query to the cluster
+7. Redirecting the client to the picked cluster
+
+The internal details of this queueing service are still under development, and this document focuses on the changes required to Presto for such a service to work. If there is interest from the community, we can provide an open-source reference implementation of this queueing service, once mature. 
```

Comment:
+1 we at uber will also be interested. What language will this be implemented in? Will it be similar to https://github.com/trinodb/trino-gateway?

Reply:
This will be implemented in Java. It is similar in some ways, but Trino Gateway is only a routing layer, this queueing service will also include queueing and scheduling. We would likely expose an interface to plug in any custom routing logic.

Reply:
It seems like this is routing + queueing.  We already have a module for the routing component, that's the `presto-router`.  Could the reference implementation be added there--i.e., could the Presto Router be configured to also take on queueing?

Reply:
Also, I read this as, a separate internal implementation will be developed, but also a reference implementation will separately be developed.  i.e. there will be two implementations.  Could we all work off of the reference implementation, and it can be configured through an SPI to integrate with internal components, rather than have two separate implementations?

Reply:
@tdcmeehan We have to see how the queueing implementation evolves, so this is TBD. We expect to use internal components and its not clear how easily we can configure these through an SPI, while still maintaining usefulness. Two integration points that could make sense are an SPI for ResourceGroupManager and for the router. The reference implementation could use the presto-router. We can discuss these points when the design is more mature. For this RFC, I want to focus on the changes required to integrate with this service. Mostly the work here for now is straightforward refactoring.